dll/win32/shell32/wine/shellpath.c

```diff
@@ -673,81 +732,89 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
     return ret;
 }
 
-BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
+BOOL WINAPI PathResolveW(_Inout_ LPWSTR path, _Inout_opt_ LPCWSTR *dirs, _In_ DWORD flags)
 {
-    DWORD dwWhich = ((flags & PRF_DONTFINDLNK) ? (WHICH_DEFAULT & ~WHICH_LNK) : WHICH_DEFAULT);
+    DWORD dwWhich = WHICH_DEFAULT; /* The extensions to be searched */
 
-    TRACE("PathResolveW(%s,%p,0x%08x)\n", debugstr_w(path), dirs, flags);
+    TRACE("(%s,%p,0x%08x)\n", debugstr_w(path), dirs, flags);
+
+    if (flags & PRF_DONTFINDLNK)
+        dwWhich &= ~WHICH_LNK; /* Don't search '.LNK' (shortcut) */
 
     if (flags & PRF_VERIFYEXISTS)
-        SetLastError(ERROR_FILE_NOT_FOUND);
+        SetLastError(ERROR_FILE_NOT_FOUND); /* We set this error code at first in verification */
 
     PathUnquoteSpacesW(path);
 
-    if (PathIsRootW(path))
+    if (PathIsRootW(path)) /* Root path */
     {
-        if ((path[0] == L'\\' && path[1] == 0) ||
-            PathIsUNCServerW(path) || PathIsUNCServerShareW(path))
-        {
-            if (flags & PRF_FIRSTDIRDEF)
-                PathQualifyExW(path, dirs[0], 0);
-            else
-                PathQualifyExW(path, NULL, 0);
-        }
+        if (path[0] == L'\\' && path[1] == UNICODE_NULL) /* '\' only? */
+            PathQualifyExW(path, ((flags & PRF_FIRSTDIRDEF) ? *dirs : NULL), 0); /* Qualify */
 
         if (flags & PRF_VERIFYEXISTS)
-            return PathFileExistsAndAttributesW(path, NULL);
+            return PathFileExistsAndAttributesW(path, NULL); /* Check the existence */
+
         return TRUE;
     }
-    else if (PathIsFileSpecW(path))
+
+    if (PathIsFileSpecW(path)) /* Filename only */
     {
-        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))
-            return TRUE;
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && /* Try to find a program? */
+            PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich)) /* Search it */
+        {
+            return TRUE; /* Found */
+        }
 
-        if (PathFindOnPathW(path, dirs))
+        if (PathFindOnPathW(path, dirs)) /* Try to find the filename in the directories */
         {
-#if (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
+#if (_WIN32_WINNT >= _WIN32_WINNT_WS03)
             if (!(flags & PRF_REQUIREABSOLUTE))
                 return TRUE;
 
             if (!PathIsAbsoluteW(path))
-                return PathMakeAbsoluteW(path) && PathFileExistsAndAttributesW(path, NULL);
+            {
+                if (!PathMakeAbsoluteW(path))
+                    return FALSE;
+                return PathFileExistsAndAttributesW(path, NULL);
+            }
```

Comment:
This #if-block could be made to look like the other one:
```suggestion
            if ((flags & PRF_REQUIREABSOLUTE) && !PathIsAbsoluteW(path))
            {
                if (!PathMakeAbsoluteW(path))
                    return FALSE;
                return PathFileExistsAndAttributesW(path, NULL);
            }
```
and replacing the #else below by #endif (and removing the old #endif)